pkg/querier/frontend/frontend.go

```diff
@@ -250,33 +250,75 @@ func (f *Frontend) RoundTripGRPC(ctx context.Context, req *ProcessRequest) (*Pro
 
 // Process allows backends to pull requests from the frontend.
 func (f *Frontend) Process(server Frontend_ProcessServer) error {
-
-	// If this request is canceled, ping the condition to unblock. This is done
-	// once, here (instead of in getNextRequest) as we expect calls to Process to
-	// process many requests.
+	var (
+		sendChan = make(chan *ProcessRequest)
+		recvChan = make(chan *ProcessResponse)
+
+		// Need buffer of 2 so goroutines reading/writing to stream don't hang
+		// around when stream dies.
+		errChan = make(chan error, 2)
+	)
+
+	// If the stream from the querier is canceled, ping the condition to unblock.
+	// This is done once, here (instead of in getNextRequest) as we expect calls
+	// to Process to process many requests.
 	go func() {
 		<-server.Context().Done()
 		f.cond.Broadcast()
 	}()
 
+	// Use a pair of goroutines to read/write from the stream and send to channels,
+	// so we can use selects to also wait on the cancellation of the request context.
+	// These goroutines will error out when the stream returns.
+	go func() {
+		for {
+			req := <-sendChan
+			err := server.Send(req)
+			if err != nil {
+				errChan <- err
+				return
+			}
+		}
+	}()
+
+	go func() {
+		for {
+			resp, err := server.Recv()
+			if err == nil {
+				recvChan <- resp
+			} else {
+				errChan <- err
+				return
+			}
+		}
+	}()
+
 	for {
 		request, err := f.getNextRequest(server.Context())
 		if err != nil {
 			return err
 		}
 
-		if err := server.Send(request.request); err != nil {
+		originalCtx := request.originalCtx
+
+		select {
+		case sendChan <- request.request:
+		case err := <-errChan:
 			request.err <- err
 			return err
+		case <-originalCtx.Done():
+			return originalCtx.Err()
 		}
 
-		response, err := server.Recv()
-		if err != nil {
+		select {
+		case resp := <-recvChan:
```

Comment:
If we don't drain this the goroutine above will be stuck and we will have a goroutine leak.

Reply:
Yup infact both goroutines will be stuck if we don't drain the `errChan`

Reply:
`errChan` is buffered so they wont leak that way; but I think they can leak as recvChan isn't drained.